### PR TITLE
Experimental: NPS prompt

### PR DESCRIPTION
**Pull request by paolodamico**
## Changes

Implements an experimental version to ask for NPS-like feedback as described here, https://github.com/PostHog/product-internal/issues/9. For now, we only ask the question to each user once, whether they reply in full, reply partially or discard altogether. We can rethink this in the future.

image

This is also a neat example of using feature flags & cohorts for data persistence and application-like logic. For now, we ask the question to users who have "discovered a learning" in the last day, but with #4349 we'll be able to do more optimized selection (e.g. users who have discovered 5 learnings or more in the last week).

**Feature flag definition**

**Current eligibility**

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious

### CODE REVIEW

**Comment by paolodamico** [comment]
@kpthatsme let me know if you want to help set up the dashboard to track submissions on this. Obviously I can do it, but would be helpful if you can. Regardless of dashboards to track overall score and submissions, I would like us to have biweekly checkpoints to review qualitative feedback. We could try word clouds to get a less bias input.

**Review by mariusandra — APPROVED** [review]
I think this looks good and technically behaves well enough.

However I went ahead, [stepped on toes](https://posthog.com/handbook/company/values#step-on-toes), and [refactored](https://github.com/PostHog/posthog/pull/4562/commits/991067ba131063e84aa2917372aa831dffdec78f) the logic and the component, moving all data into the logic, and having the component act dumb. Even made the file 7 lines shorter in the process.

This ties in to the [data-first mindset](https://kea.js.org/blog/2021/05/14/data-first-frontend-revolution/) we should adopt. If we really want to avoid subtle UX bugs, this is the approach we need to take. Having two places with state (component and logic) leads to weird inconsistencies.

(This state management solution we use happens to be Kea now, but I'd say the same thing if we'd use Mobx, Easy Peasy, Recoil, XState, Zustrand or any of the other libraries)

**Comment by mariusandra** [comment]
One more thing, I added a fix for 320px screens. This is the minimum resolution we must support in all scenes of the app (most phones are 360px, but the iPhone SE [is 320px](https://ux.stackexchange.com/questions/129084/is-320px-wide-designs-still-relevant-responsive-web) for example).

File: frontend/src/lib/experimental/NPSPrompt.tsx [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    },
+    listeners: ({ values, actions, cache }) => ({
+        stepBack: () => {
+            if (values.step === 1) {
+                actions.setPayload(null)
+            }
+        },
+        submit: () => {
+            // `nps_2106` is used to identify users who have replied to the NPS survey (via cohorts)
+            const result = ['dismissed', 'partial', 'completed'][values.step]
```
Comment (paolodamico) [inline frontend/src/lib/experimental/NPSPrompt.tsx:66]: Minor bug here: we have 4 steps, might need to repeat` 'partial'` or handle steps 1,2 differently.

Comment (mariusandra) [reply frontend/src/lib/experimental/NPSPrompt.tsx]: ah, yes :)

**Review by kpthatsme — COMMENTED** [review]
Hey @paolodamico a few thoughts -

- Earlier I think we discussed adding a time component to this i.e. you can't see this survey unless you've had a PostHog account for 45 days – I think enforcing a duration of time is important. I also think it would be valuable to learn from anyone that is active in PostHog 45 days after they sign up for an account, not just people that discovered a learning.

From the results in the survey, we can drill down further and look at the individual users / create cohorts of users and break scores down like that. Wdyt?

**Comment by paolodamico** [comment]
Thanks for the review @mariusandra! I deliberately did this code a bit haphazardly and non-optimized as it's intended as experimental code just to get some initial metrics. It's highly likely that this code will not be kept and if it is, definitely refactored to be reusable, but appreciate the teaching opportunity for myself and the team.


Re @kpthatsme, agreed 100% on the selection criteria, though we probably have to wait for #4349 to be able to make this selection as this is done through cohorts. In the meantime, I think we can use the current criteria to start getting some data.

**Comment by kpthatsme** [comment]
I was thinking we might be able to do something in the code for this, ie
for everyone that is eligible, only show if their create date was in the
time range. Where we check the date range case in the code itself via a
logic. Wdyt?

**Comment by mariusandra** [comment]
@kpthatsme the 45 day limit makes a lot of sense.

It would be amazing to specify this "user is at least 45 days old" when creating the feature flag. However for now I added a frontend check for this.

**Comment by paolodamico** [comment]
Thanks @mariusandra, beat me to it! @kpthatsme let's see what kind of data we get from this. Was giving this some more thought and maybe something like 15 days would be better, it's long enough for users to see (or not see) value but not so long that users have dropped off already and the results are too biased.

**Pull request merged**